core/unix/signal.c

```diff
@@ -4553,20 +4576,14 @@ record_pending_signal(dcontext_t *dcontext, int sig, kernel_ucontext_t *ucxt,
          */
         if (blocked && !forged && !can_always_delay[sig] &&
             safe_is_in_fcache(dcontext, pc, xsp)) {
-            dr_signal_action_t action;
             /* cache the fragment since pclookup is expensive for coarse (i#658) */
             f = fragment_pclookup(dcontext, (cache_pc)sc->SC_XIP, &wrapper);
             sc_orig = *sc;
             translate_sigcontext(dcontext, ucxt, true /*shouldn't fail*/, f);
-            /* make a copy before send_signal_to_client() tweaks it */
-            sigcontext_t sc_interrupted = *sc;
```

Comment:
So the two big structs for record_pending_signal are sc_orig and sc_interrupted.  sc_interrupted was in two places: was the compiler already sharing the space for those?  Hopefully, but I've seen compilers fail to do so before.  That leaves the two sc_orig and sc_interrupted.  This PR avoids sc_interrupted space for code paths that don't send to the client immediately: those are if blocked and synchronous, or if in client gencode.  But it doesn't change the stack usage for those two cases.  I guess the two sc_orig and sc_interrupted are both needed simultaneously?  Is the sc_interrupted copy really needed?  That makes 3 copies: sc_orig==pre-translate; *sc==in the frame; sc_interrupted==translated and..err..would have to study the code.

Looking at send_signal_to_client: it says:
```C
    /* It's safe to allocate since we do not send signals that interrupt DR.
     * With priv_mcontext_t x2 that's a little big for stack alloc.
     */
    si.mcontext = heap_alloc(dcontext, sizeof(*si.mcontext) HEAPACCT(ACCT_OTHER));
    si.raw_mcontext = heap_alloc(dcontext, sizeof(*si.raw_mcontext) HEAPACCT(ACCT_OTHER));
```

Given that, we could heap-allocate this sc_interrupted?  And avoid its stack space for all paths.

Reply:
The space for multiple instances of sc_interrupted is not shared. There is a separate stack allocation for each of them. Essentially stack was allocated for up to 3x sc_interrupted structs unnecessarily. This PR limits the scope of these allocations forcing at most one of them to be active (and free the stack space immediately after use, assuming there will be no inlining).  

Surely, it might be the case that we do not need sc_interrupted in the first place and more optis are coming.

Reply:
OK, so you're suggesting moving heap-allocating sc_interrupted into a separate PR.  OK.